File: src/shared/types/context.ts
import type { AgentStopReason, ProviderTokenUsage } from './provider'
import type { ImageAttachment } from './attachment'

export const MAIN_CONTEXT_SCOPE = 'main' as const
export const CONTEXT_SCHEMA_VERSION = 1 as const

export type ContextScopeId = typeof MAIN_CONTEXT_SCOPE | `subagent:${string}`
export type LedgerEventType =
  | 'user_message'
  | 'assistant_message'
  | 'tool_result'
  | 'skill_state_updated'
  | 'turn_completed'
  | 'turn_interrupted'
  | 'resume_state_updated'
  | 'compaction_started'
  | 'compaction_completed'
  | 'compaction_failed'
  | 'history_reverted'
  | 'legacy_import_completed'

const HISTORY_EVENT_TYPES = new Set<LedgerEventType>([
  'user_message', 'assistant_message', 'tool_result', 'turn_interrupted',
  'skill_state_updated', 'resume_state_updated', 'compaction_completed', 'legacy_import_completed',
  'history_reverted'
])

export function contextScopeForSubAgent(runId: string): ContextScopeId {
  if (!runId.trim()) throw new Error('runId is required')
  return `subagent:${runId}`
}

export function eventChangesHistory(type: LedgerEventType): boolean {
  return HISTORY_EVENT_TYPES.has(type)
}

export interface GoalSnapshot {
  id?: string
  title?: string
  originalPrompt: string
  normalizedGoal?: string
  keyRequirements: string[]
  nonGoals?: string[]
  successCriteria?: string[]
  updatedAt?: string
}

export interface TaskPlan {
  currentStep: string
  completedSteps: string[]
  pendingSteps: string[]
}

export interface ResumeState {
  currentGoalId: string
  currentPhase: string
  currentStep: string
  lastCompletedStep?: string
  nextAction: string
  openQuestions: string[]
  blockedBy: string[]
  filesTouched: string[]
  filesToInspectNext: string[]
  validationPending: string[]
  validationResults?: Array<{
    commandOrCheck: string
    status: 'passed' | 'failed'
    result?: string
  }>
  goal?: GoalSnapshot
  plan?: TaskPlan
  contextFiles?: string[]
  lastTrimmedAt?: number
  updatedAt?: string
}

export interface VersionedResumeState {
  revision: number
  coveredThroughSequence: number
  source: 'explicit_tool' | 'compaction' | 'framework'
  updatedAt: string
  state: ResumeState
}

export interface NormalizedToolCall {
  id: string
  name: string
  arguments: string
  thoughtSignature?: string
}

export interface FileContextReference {
  path: string
  sha256: string
  operation: 'read' | 'edit' | 'write'
  /** True only when this message contains the file contents visible to the model. */
  contentIncluded: boolean
  /** Hash of the exact rendered content, used to deduplicate identical Read ranges. */
  contentSha256?: string
  offset?: number
  limit?: number
  /** 1-indexed character offset within a single requested line. */
  characterOffset?: number
  /** Durable ledger ordering used to restore the most recently accessed files. */
  accessSequence?: number
  /** Exact block boundaries inside a Read tool result, enabling per-file projection. */
  resultBlockStart?: number
  resultBlockEnd?: number
}

export interface PostCompactionFileBlock {
  reference: FileContextReference
  /** Numbered file text. It is JSON-escaped before entering a provider request. */
  content: string
  /** Filesystem version used to invalidate restored text after external changes. */
  statSignature: string
  realPath?: string
}

export interface PostCompactionFileContext {
  /** Safe JSON data projection; legacy snapshots may contain the old XML rendering here. */
  content: string
  fileReferences: FileContextReference[]
  blocks?: PostCompactionFileBlock[]
  createdAt: string
  sourceSequence?: number
}

export interface InvokedSkillContextEntry {
  name: string
  content: string
  invokedSequence: number
}

export type SessionSkillStatus = 'active' | 'inactive' | 'disabled'

export interface SessionSkillState {
  name: string
  status: SessionSkillStatus
  content?: string
  contentHash?: string
  args?: string
  source: 'model' | 'user' | 'recovery'
  reason?: string
  updatedAt: string
  updatedSequence: number
}

export interface PostCompactionSkillContext {
  content: string
  skills: InvokedSkillContextEntry[]
  createdAt: string
  sourceSequence?: number
}

export interface NormalizedModelMessage {
  id: string
  /** Renderer message identity used for durable conversation history operations. */
  clientMessageId?: string
  turnId: string
  role: 'user' | 'assistant' | 'tool'
  content: string
  toolCalls?: NormalizedToolCall[]
  toolCallId?: string
  name?: string
  status: 'complete' | 'interrupted'
  createdAt: string
  sourceSequence?: number
  attachments?: ImageAttachment[]
  /** Client-side file working-set metadata. Provider adapters do not serialize it. */
  fileReferences?: FileContextReference[]
}

export interface ModelContextItem {
  kind: 'system' | 'compaction_summary' | 'resume_state' | 'skill_context' | 'skill_state' | 'file_context' | 'user' | 'assistant' | 'tool'
  message: NormalizedModelMessage | {
    role: 'system'
    content: string
    fileReferences?: FileContextReference[]
    sourceSequence?: number
  }
}

export interface CompactionSummaryV1 {
  version: 1
  goal: {
    originalRequest?: string
    currentObjective: string
    requirements: string[]
    successCriteria: string[]
  }
  status: {
    phase: string
    completed: string[]
    inProgress: string[]
    nextActions: string[]
  }
  decisions: Array<{ decision: string; rationale?: string }>
  facts: Array<{ fact: string; evidence?: string }>
  files: Array<{
    path: string
    relevance: string
    state: 'read' | 'modified' | 'created' | 'deleted' | 'unknown'
  }>
  validation: Array<{
    commandOrCheck: string
    result: string
    status: 'passed' | 'failed' | 'pending'
  }>
  errors: Array<{ symptom: string; cause?: string; resolution?: string }>
  openQuestions: string[]
  userInstructions: string[]
  coveredThroughSequence: number
}

export interface CompactionSummaryV2 {
  version: 2
  format: 'text'
  content: string
  coveredThroughSequence: number
  /** Oldest source content omitted only as a last-resort Provider overflow recovery. */
  truncatedPrefixThroughSequence?: number
}

export type CompactionSummary = CompactionSummaryV1 | CompactionSummaryV2

export type ContextPressureLevel = 'normal' | 'warning' | 'prune' | 'compact' | 'overflow'
export type ContextEstimateSource = 'provider' | 'tokenizer' | 'heuristic'

export interface ContextBudgetSnapshot {
  hardInputLimit: number
  usableInputBudget: number
  systemPromptTokens: number
  toolSchemaTokens: number
  instructionTokens: number
  protocolTokens: number
  summaryTokens: number
  recentHistoryTokens: number
  rawHistoryTokens: number
  currentInputTokens: number
  outputReserveTokens: number
  safetyMarginTokens: number
  totalInputTokens: number
  providerAdjustmentTokens: number
  pressureLevel: ContextPressureLevel
  estimateSource: ContextEstimateSource
  historyVersion: number
}

export type CompactionTrigger =
  | 'auto_threshold'
  | 'provider_overflow'
  | 'manual'
  | 'model_downshift'
  | 'migration'

export type ContextErrorCode =
  | 'LEDGER_WRITE_FAILED'
  | 'LEDGER_CORRUPTED'
  | 'SNAPSHOT_COMMIT_FAILED'
  | 'BUDGET_UNAVAILABLE'
  | 'CURRENT_INPUT_TOO_LARGE'
  | 'COMPACTION_SUMMARY_FAILED'
  | 'COMPACTION_SCHEMA_INVALID'
  | 'COMPACTION_STALE_VERSION'
  | 'COMPACTION_INSUFFICIENT_REDUCTION'
  | 'PROVIDER_CONTEXT_OVERFLOW'
  | 'LEGACY_MIGRATION_FAILED'

export interface UserMessagePayload {
  message: NormalizedModelMessage
  providerId?: string
  model?: string
  commandMetadata?: unknown
}

export interface AssistantMessagePayload {
  message: NormalizedModelMessage
  usage?: ProviderTokenUsage
  /** Fingerprint of the exact Provider request measured by usage. */
  requestFingerprint?: string
}

export interface ToolResultPayload {
  message: NormalizedModelMessage
  status: 'success' | 'error' | 'interrupted'
  fullResultSha256?: string
}

export interface SkillStateUpdatedPayload {
  name: string
  status: SessionSkillStatus
  content?: string
  contentHash?: string
  args?: string
  source: SessionSkillState['source']
  reason?: string
}

export interface TurnCompletedPayload {
  stopReason: AgentStopReason
  usage?: ProviderTokenUsage
  completedAt: string
}

export interface TurnInterruptedPayload {
  reason: string
  interruptedMessages: NormalizedModelMessage[]
}

export interface CompactionStartedPayload {
  trigger: CompactionTrigger
  sourceHistoryVersion: number
  candidateThroughSequence: number
  tokensBefore: number
}

export interface ObservedProviderInputLimit {
  providerId?: string
  model?: string
  maxInputTokens: number
}

export interface CompactionCompletedPayload {
  trigger: CompactionTrigger
  sourceHistoryVersion: number
  coveredThroughSequence: number
  retainedFromSequence?: number
  tokensBefore: number
  tokensAfter: number
  sourceHash: string
  summary: CompactionSummary
  observedProviderInputLimit?: ObservedProviderInputLimit
  resumeState?: VersionedResumeState
  activeMessages: NormalizedModelMessage[]
  postCompactionFileContext?: PostCompactionFileContext
  postCompactionSkillContext?: PostCompactionSkillContext
  skillStates?: SessionSkillState[]
  postCompactionSkillStates?: SessionSkillState[]
}

export interface CompactionFailedPayload {
  trigger: CompactionTrigger
  stage: string
  code: ContextErrorCode
  message: string
  retryable: boolean
}

export interface HistoryRevertedPayload {
  sourceHistoryVersion: number
  targetUiMessageId: string
  targetMessageId: string
  activeMessages: NormalizedModelMessage[]
  skillStates?: SessionSkillState[]
}

export interface LegacyImportCompletedPayload {
  sourceHash: string
  mode: 'summary' | 'recent-text-fallback'
  activeMessages: NormalizedModelMessage[]
  summary?: CompactionSummary
}

export interface LedgerPayloadByType {
  user_message: UserMessagePayload
  assistant_message: AssistantMessagePayload
  tool_result: ToolResultPayload
  skill_state_updated: SkillStateUpdatedPayload
  turn_completed: TurnCompletedPayload
  turn_interrupted: TurnInterruptedPayload
  resume_state_updated: { resumeState: VersionedResumeState }
  compaction_started: CompactionStartedPayload
  compaction_completed: CompactionCompletedPayload
  compaction_failed: CompactionFailedPayload
  history_reverted: HistoryRevertedPayload
  legacy_import_completed: LegacyImportCompletedPayload
}

export interface LedgerEvent<
  TType extends LedgerEventType = LedgerEventType,
  TPayload = LedgerPayloadByType[TType]
> {
  schemaVersion: 1
  eventId: string
  sessionId: string
  contextScopeId: ContextScopeId
  sequence: number
  historyVersion: number
  turnId?: string
  createdAt: string
  type: TType
  payload: TPayload
}

export type AnyLedgerEvent = {
  [TType in LedgerEventType]: LedgerEvent<TType, LedgerPayloadByType[TType]>
}[LedgerEventType]

export interface SessionRuntimeScopeSnapshot {
  historyVersion: number
  activeMessages: NormalizedModelMessage[]
  latestCompaction?: CompactionSummary
  observedProviderInputLimit?: ObservedProviderInputLimit
  resumeState?: VersionedResumeState
  lastCompletedTurnId?: string
  lastInterruptedTurnId?: string
  legacyImport?: {
    sourceHash: string
    mode: 'summary' | 'recent-text-fallback'
    eventId: string
  }
  latestCompactionResumeRevision?: number
  lastProviderId?: string
  lastModel?: string
  lastProviderUsage?: ProviderTokenUsage
  lastProviderUsageMessageId?: string
  lastProviderUsageProviderId?: string
  lastProviderUsageModel?: string
  lastProviderUsageRequestFingerprint?: string
  postCompactionFileContext?: PostCompactionFileContext
  postCompactionSkillContext?: PostCompactionSkillContext
  skillStates?: SessionSkillState[]
  postCompactionSkillStates?: SessionSkillState[]
}

export interface SessionRuntimeSnapshot {
  schemaVersion: 1
  sessionId: string
  throughSequence: number
  createdAt: string
  scopes: Record<string, SessionRuntimeScopeSnapshot>
}

export interface SessionRuntimeRef {
  schemaVersion: 2
  ledgerVersion: 1
  migratedAt?: string
  legacySourceHash?: string
  legacyImportMode?: 'summary' | 'recent-text-fallback'
}

export interface StreamRequestV2 {
  streamId: string
  providerId: string
  model: string
  sessionId: string
  input: {
    text: string
    attachments?: ImageAttachment[]
    isSystem?: boolean
    commandMetadata?: unknown
  }
}

File: crates/codez-core/src/context.rs
use std::{borrow::Cow, collections::HashMap, fmt};

use serde::{Deserialize, Deserializer, Serialize, Serializer, de::Error as _};
use thiserror::Error;

use crate::{
    ComposerImageAttachment,
    provider::{AgentStopReason, ProviderTokenUsage},
};

pub const MAIN_CONTEXT_SCOPE: &str = "main";
pub const CONTEXT_SCHEMA_VERSION: u16 = 1;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ContextScopeId {
    Main,
    Subagent(String),
}

#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("context scope must be 'main' or a non-empty 'subagent:<id>' value")]
pub struct ContextScopeIdError;

impl ContextScopeId {
    pub fn parse(value: &str) -> Result<Self, ContextScopeIdError> {
        if value == MAIN_CONTEXT_SCOPE {
            return Ok(Self::Main);
        }
        let Some(identifier) = value.strip_prefix("subagent:") else {
            return Err(ContextScopeIdError);
        };
        if identifier.is_empty()
            || identifier.len() > 160
            || identifier.chars().any(char::is_control)
        {
            return Err(ContextScopeIdError);
        }
        Ok(Self::Subagent(identifier.to_string()))
    }

    #[must_use]
    pub fn as_key(&self) -> Cow<'_, str> {
        match self {
            Self::Main => Cow::Borrowed(MAIN_CONTEXT_SCOPE),
            Self::Subagent(identifier) => Cow::Owned(format!("subagent:{identifier}")),
        }
    }
}

impl fmt::Display for ContextScopeId {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.as_key())
    }
}

impl Serialize for ContextScopeId {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.collect_str(self)
    }
}

impl<'de> Deserialize<'de> for ContextScopeId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = String::deserialize(deserializer)?;
        Self::parse(&value).map_err(D::Error::custom)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LedgerEventType {
    UserMessage,
    AssistantMessage,
    ToolResult,
    SkillStateUpdated,
    TurnCompleted,
    TurnInterrupted,
    ResumeStateUpdated,
    CompactionStarted,
    CompactionCompleted,
    CompactionFailed,
    HistoryReverted,
    LegacyImportCompleted,
}

impl LedgerEventType {
    #[must_use]
    pub const fn changes_history(self) -> bool {
        matches!(
            self,
            Self::UserMessage
                | Self::AssistantMessage
                | Self::ToolResult
                | Self::SkillStateUpdated
                | Self::TurnInterrupted
                | Self::ResumeStateUpdated
                | Self::CompactionCompleted
                | Self::HistoryReverted
                | Self::LegacyImportCompleted
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GoalSnapshot {
    pub id: Option<String>,
    pub title: Option<String>,
    pub original_prompt: String,
    pub normalized_goal: Option<String>,
    pub key_requirements: Vec<String>,
    pub non_goals: Option<Vec<String>>,
    pub success_criteria: Option<Vec<String>>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskPlan {
    pub current_step: String,
    pub completed_steps: Vec<String>,
    pub pending_steps: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResumeState {
    pub current_goal_id: String,
    pub current_phase: String,
    pub current_step: String,
    pub last_completed_step: Option<String>,
    pub next_action: String,
    pub open_questions: Vec<String>,
    pub blocked_by: Vec<String>,
    pub files_touched: Vec<String>,
    pub files_to_inspect_next: Vec<String>,
    pub validation_pending: Vec<String>,
    pub validation_results: Option<Vec<ValidationResult>>,
    pub goal: Option<GoalSnapshot>,
    pub plan: Option<TaskPlan>,
    pub context_files: Option<Vec<String>>,
    pub last_trimmed_at: Option<u64>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ValidationResult {
    pub command_or_check: String,
    pub status: String,
    pub result: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VersionedResumeState {
    pub revision: u32,
    pub covered_through_sequence: u32,
    pub source: String,
    pub updated_at: String,
    pub state: ResumeState,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NormalizedToolCall {
    pub id: String,
    pub name: String,
    pub arguments: String,
    pub thought_signature: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileContextReference {
    pub path: String,
    pub sha256: String,
    pub operation: String,
    pub content_included: bool,
    pub content_sha256: Option<String>,
    pub offset: Option<u64>,
    pub limit: Option<u64>,
    pub character_offset: Option<u64>,
    pub access_sequence: Option<u64>,
    pub result_block_start: Option<u32>,
    pub result_block_end: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PostCompactionFileBlock {
    pub reference: FileContextReference,
    pub content: String,
    pub stat_signature: String,
    pub real_path: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PostCompactionFileContext {
    pub content: String,
    pub file_references: Vec<FileContextReference>,
    pub blocks: Option<Vec<PostCompactionFileBlock>>,
    pub created_at: String,
    pub source_sequence: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InvokedSkillContextEntry {
    pub name: String,
    pub content: String,
    pub invoked_sequence: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSkillState {
    pub name: String,
    pub status: String,
    pub content: Option<String>,
    pub content_hash: Option<String>,
    pub args: Option<String>,
    pub source: String,
    pub reason: Option<String>,
    pub updated_at: String,
    pub updated_sequence: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PostCompactionSkillContext {
    pub content: String,
    pub skills: Vec<InvokedSkillContextEntry>,
    pub created_at: String,
    pub source_sequence: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NormalizedModelMessage {
    pub id: String,
    pub client_message_id: Option<String>,
    pub turn_id: String,
    pub role: String,
    pub content: String,
    pub tool_calls: Option<Vec<NormalizedToolCall>>,
    pub tool_call_id: Option<String>,
    pub name: Option<String>,
    pub status: String,
    pub created_at: String,
    pub source_sequence: Option<u32>,
    pub attachments: Option<Vec<ComposerImageAttachment>>,
    pub file_references: Option<Vec<FileContextReference>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LedgerEvent {
    pub schema_version: u16,
    pub event_id: String,
    pub session_id: String,
    pub context_scope_id: ContextScopeId,
    pub sequence: u32,
    pub history_version: u32,
    pub turn_id: Option<String>,
    pub created_at: String,
    pub r#type: LedgerEventType,
    pub payload: serde_json::Value,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LedgerAppendRequest {
    pub event_id: String,
    pub session_id: String,
    pub context_scope_id: ContextScopeId,
    pub turn_id: Option<String>,
    pub created_at: String,
    pub r#type: LedgerEventType,
    pub payload: serde_json::Value,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionRuntimeScopeSnapshot {
    pub history_version: u32,
    pub active_messages: Vec<NormalizedModelMessage>,
    pub latest_compaction: Option<serde_json::Value>,
    pub observed_provider_input_limit: Option<serde_json::Value>,
    pub resume_state: Option<VersionedResumeState>,
    pub last_completed_turn_id: Option<String>,
    pub last_interrupted_turn_id: Option<String>,
    pub legacy_import: Option<serde_json::Value>,
    pub latest_compaction_resume_revision: Option<u32>,
    pub last_provider_id: Option<String>,
    pub last_model: Option<String>,
    pub last_provider_usage: Option<ProviderTokenUsage>,
    pub last_provider_usage_message_id: Option<String>,
    pub last_provider_usage_provider_id: Option<String>,
    pub last_provider_usage_model: Option<String>,
    pub last_provider_usage_request_fingerprint: Option<String>,
    pub post_compaction_file_context: Option<PostCompactionFileContext>,
    pub post_compaction_skill_context: Option<PostCompactionSkillContext>,
    pub skill_states: Option<Vec<SessionSkillState>>,
    pub post_compaction_skill_states: Option<Vec<SessionSkillState>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionRuntimeSnapshot {
    pub schema_version: u16,
    pub session_id: String,
    pub through_sequence: u32,
    pub created_at: String,
    pub scopes: HashMap<String, SessionRuntimeScopeSnapshot>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ModelContextItemMessage {
    Normalized(Box<NormalizedModelMessage>),
    System {
        role: String,
        content: String,
        file_references: Option<Vec<FileContextReference>>,
        source_sequence: Option<u32>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelContextItem {
    pub kind: String,
    pub message: ModelContextItemMessage,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UserMessagePayload {
    pub message: NormalizedModelMessage,
    pub provider_id: Option<String>,
    pub model: Option<String>,
    pub command_metadata: Option<serde_json::Value>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AssistantMessagePayload {
    pub message: NormalizedModelMessage,
    pub usage: Option<ProviderTokenUsage>,
    pub request_fingerprint: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolResultPayload {
    pub message: NormalizedModelMessage,
    pub status: String,
    pub full_result_sha256: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillStateUpdatedPayload {
    pub name: String,
    pub status: String,
    pub content: Option<String>,
    pub content_hash: Option<String>,
    pub args: Option<String>,
    pub source: String,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnCompletedPayload {
    pub stop_reason: AgentStopReason,
    pub usage: Option<ProviderTokenUsage>,
    pub completed_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnInterruptedPayload {
    pub reason: String,
    pub interrupted_messages: Vec<NormalizedModelMessage>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResumeStateUpdatedPayload {
    pub resume_state: VersionedResumeState,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompactionCompletedPayload {
    pub trigger: String,
    pub source_history_version: u32,
    pub covered_through_sequence: u32,
    pub retained_from_sequence: Option<u32>,
    pub tokens_before: u32,
    pub tokens_after: u32,
    pub source_hash: String,
    pub summary: serde_json::Value,
    pub observed_provider_input_limit: Option<serde_json::Value>,
    pub resume_state: Option<VersionedResumeState>,
    pub active_messages: Vec<NormalizedModelMessage>,
    pub post_compaction_file_context: Option<PostCompactionFileContext>,
    pub post_compaction_skill_context: Option<PostCompactionSkillContext>,
    pub skill_states: Option<Vec<SessionSkillState>>,
    pub post_compaction_skill_states: Option<Vec<SessionSkillState>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompactionStartedPayload {
    pub trigger: String,
    pub source_history_version: u32,
    pub candidate_through_sequence: u32,
    pub tokens_before: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompactionFailedPayload {
    pub trigger: String,
    pub stage: String,
    pub code: String,
    pub message: String,
    pub retryable: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryRevertedPayload {
    pub source_history_version: u32,
    pub target_ui_message_id: String,
    pub target_message_id: String,
    pub active_messages: Vec<NormalizedModelMessage>,
    pub skill_states: Option<Vec<SessionSkillState>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LegacyImportCompletedPayload {
    pub source_hash: String,
    pub mode: String,
    pub active_messages: Vec<NormalizedModelMessage>,
    pub summary: Option<serde_json::Value>,
}

#[cfg(test)]
mod tests {
    use super::{ContextScopeId, LedgerEventType};

    #[test]
    fn context_scope_uses_the_stable_string_wire_format() {
        let serialized = serde_json::to_string(&ContextScopeId::Subagent("run-7".to_string()))
            .expect("fixed scope must serialize");

        assert_eq!(serialized, r#""subagent:run-7""#);
    }

    #[test]
    fn only_model_visible_events_advance_history() {
        assert!(LedgerEventType::UserMessage.changes_history());
        assert!(!LedgerEventType::TurnCompleted.changes_history());
    }
}
